usdt-impl/src/stapsdt/args.rs

```diff
@@ -0,0 +1,239 @@
+    sign: Sign::Unsigned,
+    width: BitWidth::Bit64,
+};
+
+/// Convert a type and register index to its GNU Assembler operation as a
+/// String.
+fn native_data_type_to_asm_op(typ: &NativeDataType, reg_index: u8) -> String {
+    match typ {
+        NativeDataType::Integer(int) => integer_to_asm_op(int, reg_index).into(),
+        // Integer pointers are dereferenced by wrapping the pointer assembly
+        // into parentheses.
+        NativeDataType::Pointer(_) => format!("({})", integer_to_asm_op(&POINTER, reg_index)),
+        NativeDataType::String => integer_to_asm_op(&POINTER, reg_index).into(),
+    }
+}
+
+/// Convert a type to its GNU Assembler size representation as a string.
+fn native_data_type_to_arg_size(typ: &NativeDataType) -> &'static str {
+    match typ {
+        NativeDataType::Integer(int) => integer_to_arg_size(int),
+        NativeDataType::Pointer(_) | NativeDataType::String => integer_to_arg_size(&POINTER),
+        // Note: If NativeDataType::Float becomes supported, it will need an
+        // "f" suffix in the type, eg. `4f` or `8f`.
+    }
+}
+
+/// Convert a DataType and register index to its GNU Assembler operation as a
+/// String.
+fn data_type_to_asm_op(typ: &DataType, reg_index: u8) -> String {
+    match typ {
+        DataType::Native(ty) => native_data_type_to_asm_op(ty, reg_index),
+        DataType::UniqueId => integer_to_asm_op(&UNIQUE_ID, reg_index).into(),
+        DataType::Serializable(_) => integer_to_asm_op(&POINTER, reg_index).into(),
+    }
+}
+
+/// Convert a DataType to its STAPSDT probe argument size representation as a
+/// String.
+fn data_type_to_arg_size(typ: &DataType) -> &'static str {
+    match typ {
+        DataType::Native(ty) => native_data_type_to_arg_size(ty),
+        DataType::UniqueId => integer_to_arg_size(&UNIQUE_ID),
+        DataType::Serializable(_) => integer_to_arg_size(&POINTER),
+    }
+}
+
+/// ## Format a STAPSDT probe argument into the SystemTap argument format.
+/// Source: https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation
+///
+/// ### Summary
+///
+/// Argument format is `Nf@OP`, N is an optional `-` to signal signedness
+/// followed by one of `{1,2,4,8}` for bit width, `f` is an optional marker
+/// for floating point values, @ is a separator, and OP is the
+/// "actual assembly operand". The assembly operand is given in the GNU
+/// Assembler format. See
+/// https://en.wikibooks.org/wiki/X86_Assembly/GNU_assembly_syntax
+/// for details.
+///
+/// ### Examples
+///
+/// 1. Read a u64 from RDI: `8@%rdi`.
+/// 2. Read an i32 through a pointer in RSI: `-4@(%rsi)`.
+/// 3. Read an f64 through a pointer in RDI: `8f@(%rdi)`.
+///    (Not sure if `-` should be added.)
+/// 4. Read a u64 through a pointer with an offset: `8%-4(%rdi)`.
+pub(crate) fn format_argument((reg_index, typ): (usize, &DataType)) -> String {
+    format!(
+        "{}@{}",
+        data_type_to_arg_size(typ),
+        data_type_to_asm_op(typ, reg_index as u8)
```

Comment:
I'd prefer `u8::try_from(reg_index).unwrap()` here. It's unlikely we'll have > 256 arguments, but I prefer the safety anyway.

Reply:
Good point, fixed <3